File: 2. Simple Linear Regression/simple_linear_regression.R
library(caTools)
library(ggplot2)
this.dir <- dirname(parent.frame(2)$ofile)
setwd(this.dir)
dataset = read.csv("Salary_Data.csv")

set.seed(123)

# Splitting to test and train
split = sample.split(dataset$Salary, SplitRatio = 2 / 3)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

#fitting line
regressor = lm(formula = Salary ~ YearsExperience, data = training_set)

# Predicting line
y_pred = predict(regressor, newdata = training_set)

pl = ggplot() +
  geom_point(aes(x = test_set$YearsExperience, y = test_set$Salary)) + 
  geom_line(aes(x = training_set$YearsExperience, y = y_pred)) +
  ggtitle("Salary and experience") +
  xlab("Salary") +
  ylab("Years of experience") +
  theme_bw()

print(pl)

File: 2. Simple Linear Regression/simple_linear_regression.py
import numpy as np 
import matplotlib.pyplot as plt 
import pandas as pd 
import os
import pprint
from tabulate import tabulate

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Setting working directory to the directory of this script
abspath = os.path.abspath(__file__)
dname = os.path.dirname(abspath)
os.chdir(dname)

# Reading data
dataset = pd.read_csv("Salary_Data.csv")
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 1].values

#splitting test and train dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

# Fitting simple linear regression -- least squares
regressor = LinearRegression() # simple least squares linear regression
regressor.fit(X_train, y_train)
y_hat = regressor.predict(X_test)


pretty_matrix = np.vstack((X_test.T, y_test, y_hat, 100 * (y_hat - y_test)/y_test) ).T
print(tabulate(pretty_matrix, headers=["Experience", "Actual salary", "Predicted salary", "Diff"]))

# Visualising results
plt.scatter(X, y, color='red')
plt.plot(X_test, y_hat, color = "blue")
plt.title("Salary and experience, training set")
plt.xlabel("years of experience")
plt.ylabel("Salary")
plt.show()
